Fix Array ctor integer widening and add Reflection tests

AaronRobinsonMSFT opened the pull request:
The Array ctor path via Reflection had a bug in its integer widening, so converted that to the correct path.
Add tests for Reflection Invoke where the input `object[]` is changed during invocation.
Remove some coreclr code in the Invoke path.

AaronRobinsonMSFT committed:
coreclr/

Remove always false IsStructRequiringStackAllocRetBuf().
Extract the conversion of array constructor arguments into managed.
Expose new FCall for Array ctor call.

AaronRobinsonMSFT committed:
libraries/

Add special case for array construction during Invoke.

AaronRobinsonMSFT committed:
mono/

Create new icall for Array ctor call.
Create macro define for SPAN_T.

AaronRobinsonMSFT committed:
mono/

Build issues.

vargaz commented:
The mono changes look ok.

AaronRobinsonMSFT commented:
/cc @steveharter @buyaa-n @elinor-fung @jkotas

jkotas commented on src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.Mono.cs:
```diff
@@ -869,19 +869,28 @@ private void InvokeClassConstructor()
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal object InvokeArrayCtorWorker(in Span<int> arguments)
```
The general guidance is to avoid passing `Span` as `in`. I understand that you need it for the FCall, but the rest of the code should follow the general guidance.

jkotas replied:
For example, local variable passed as `in` makes the local address-taken that tends to produce less efficient code. Using `in` for Span is a very complex trade-off.

AaronRobinsonMSFT replied:
Grr. Boo.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.cs:
```diff
@@ -158,14 +158,25 @@ public override object Invoke(BindingFlags invokeAttr, Binder? binder, object?[]
+            // Array construction is a special case.
+            if (DeclaringType is not null && DeclaringType.IsArray)
+            {
+                Span<int> args = stackalloc int[actualCount];
```
Is adding `stackalloc` here going to visibly regress the non-array path through the method?

AaronRobinsonMSFT replied:
At this point it is simply a call out to a dedicated FCall and I observe no impact.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.cs:
```diff
@@ -158,14 +158,25 @@ public override object Invoke(BindingFlags invokeAttr, Binder? binder, object?[]
+                Span<int> args = stackalloc int[actualCount];
+                for (int i = 0; i < actualCount; ++i)
+                {
+                    args[i] = (int)parameters![i]!;
```
The existing implementation does implicit widening, like all other reflection invoke APIs. For example, it is possible to reflection invoke the method with `byte` arguments that will be widened to `int`. This does not do the implicit widening anymore.

AaronRobinsonMSFT replied:
Good call. Let me see what I can do here.

AaronRobinsonMSFT replied:
This proved to not be a good use of my time. I am simply moving some of this into a separate code path to declutter the most common `Invoke` code path.

AaronRobinsonMSFT committed:
coreclr/

AaronRobinsonMSFT committed:
mono/

AaronRobinsonMSFT committed:
libraries/

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.cs:
```diff
@@ -158,6 +158,12 @@ public override object Invoke(BindingFlags invokeAttr, Binder? binder, object?[]
+            // Array construction is a special case.
+            if (DeclaringType is not null && DeclaringType.IsArray)
+            {
+                return InvokeArrayCtorWorker(parameters!);
```
We will skip `CheckArguments` below for the array constructor now. Is `CheckArguments` guaranteed to be no-op for the array constructors? It is not obvious to me that it is the case.

AaronRobinsonMSFT replied:
I believe it is based on what is expected in the new `InvokeArrayCtor` which was previously `InvokeArrayConstructor` and has the following argument processing&mdash;all arguments must be primitives and "widenable" to `I4`. I don't see the logic in `CheckArguments` that would needed to run for this case.

```cpp
        if (!objs->GetAt(i))
            COMPlusThrowArgumentException(W("parameters"), W("Arg_NullIndex"));

        MethodTable* pMT = objs->GetAt(i)->GetMethodTable();
        CorElementType oType = TypeHandle(pMT).GetVerifierCorElementType();

        if (!InvokeUtil::IsPrimitiveType(oType) || !InvokeUtil::CanPrimitiveWiden(ELEMENT_TYPE_I4,oType))
            COMPlusThrow(kArgumentException,W("Arg_PrimWiden"));
```

jkotas replied:
This prints "2" today. Is it still work after this change?
```csharp
using System.Globalization;
using System.Reflection;

var ctor = typeof(int[,]).GetConstructor(new[] { typeof(int), typeof(int) });
var arr = ctor.Invoke(BindingFlags.Default, new MyBinder(), new object[] { "1", "2" }, null);
Console.WriteLine(((int[,])arr).Length);

class MyBinder : Binder
{
    public override FieldInfo BindToField(BindingFlags bindingAttr, FieldInfo[] match, object value, CultureInfo? culture)
        => throw new NotImplementedException();

    public override MethodBase BindToMethod(BindingFlags bindingAttr, MethodBase[] match, ref object?[] args, ParameterModifier[]? modifiers, CultureInfo? culture, string[]? names, out object? state)
        => throw new NotImplementedException();

    public override object ChangeType(object value, Type type, CultureInfo? culture)
        => int.Parse((string)value);

    public override void ReorderArgumentArray(ref object?[] args, object state)
        => throw new NotImplementedException();

    public override MethodBase? SelectMethod(BindingFlags bindingAttr, MethodBase[] match, Type[] types, ParameterModifier[]? modifiers)
        => throw new NotImplementedException();

    public override PropertyInfo? SelectProperty(BindingFlags bindingAttr, PropertyInfo[] match, Type? returnType, Type[]? indexes, ParameterModifier[]? modifiers)
        => throw new NotImplementedException();
}
```

AaronRobinsonMSFT replied:
```csharp
        public override object ChangeType(object value, Type type, CultureInfo? culture)
            => int.Parse((string)value);
```

-.-    That is evil...

AaronRobinsonMSFT replied:
I am adding a test case too.

AaronRobinsonMSFT committed:
libraries/

Add tests for Reflection Binder type conversion support during Invoke.

jkotas commented on src/coreclr/vm/reflectioninvocation.cpp:
```diff
@@ -1166,8 +1095,61 @@ FCIMPL5(Object*, RuntimeMethodHandle::InvokeMethod,
+        if (!InvokeUtil::IsPrimitiveType(oType) || !InvokeUtil::CanPrimitiveWiden(ELEMENT_TYPE_I4,oType))
+            COMPlusThrow(kArgumentException,W("Arg_PrimWiden"));
+
+        memcpy(&indexes[i], objs->GetAt(i)->UnBox(), pMT->GetNumInstanceFieldBytes());
```
I know that this was here before, but it is wrong. The widening has to do proper sign-extension like what the regular invoke does (regular PInvoke does it using `InvokeUtil::CopyArg` method).

For example:
```csharp
using System.Reflection;

var ctor = typeof(int[,]).GetConstructor(new[] { typeof(int), typeof(int), typeof(int), typeof(int) });
var arr = (int[,])ctor.Invoke(new object[] { (sbyte)-1, (sbyte)2, (sbyte)-3, (sbyte)4 });
Console.WriteLine(arr.GetLowerBound(0));
Console.WriteLine(arr.GetLowerBound(1));
```
prints 255 253, but it should print -1 -3 instead.

AaronRobinsonMSFT replied:
This is a bit clunky. It would appear the `InvokeUtil::CopyArg` and `InvokeUtil::CreatePrimitiveValue` both have bugs in them. They indiscriminately copy everything into the `ARG_SLOT` which for this case isn't the size of the destination. We are using a stack allocation and since `ARG_SLOT` is always 64-bits it corrupts the stack. I've fixed `InvokeUtil::CreatePrimitiveValue` to do the right thing.

jkotas replied:
`ARG_SLOT` is what regular method invoke arguments get normalized into so I think it makes sense that `InvokeUtil::CopyArg` and `InvokeUtil::CreatePrimitiveValue` expect the destination to be `ARG_SLOT`.

AaronRobinsonMSFT replied:
Right. I missed how this API was being used. I found an example of the correct way to use this and am reverting my changes to it.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.cs:
```diff
@@ -165,6 +165,12 @@ public override object Invoke(BindingFlags invokeAttr, Binder? binder, object?[]
                 arguments = CheckArguments(ref stackArgs, parameters, binder, invokeAttr, culture, ArgumentTypes);
```
Just noticed unrelated bug above: `if ((InvocationFlags & InvocationFlags.RunClassConstructor) != 0)` condition above looks misplaced. It should be after we validate the number of arguments. We should throw exception when somebody tries to invoke class constructor with arguments.

jkotas commented on src/coreclr/vm/reflectioninvocation.cpp:
```diff
@@ -1166,8 +1095,61 @@ FCIMPL5(Object*, RuntimeMethodHandle::InvokeMethod,
+        OBJECTREF retVal;
+    } gc;
+
+    gc.pSig = (SIGNATURENATIVEREF)pSigUNSAFE;
```
The old code also had a check for `IsSharedByGenericInstantiations`. We may need it here as well.

(In theory, it should not possible to get a managed Type for `IsSharedByGenericInstantiations` type. In practice, we always had odd corner cases or bugs how one can end up with them.)

AaronRobinsonMSFT replied:
Yes. I did check that from the typical case and it trigger long before it got here. I can add it back here if you think it is warranted.

jkotas replied:
I would either add it here or delete it from the other place so that things are consistent.

AaronRobinsonMSFT committed:
coreclr/

Make sure integral types respect sign extension during widen operation.
Fix InvokeUtil::CreatePrimitiveValue to write destination size.

AaronRobinsonMSFT committed:
libraries/

Validate arg count prior to running static constructor.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.cs:
```diff
@@ -165,6 +165,12 @@ public override object Invoke(BindingFlags invokeAttr, Binder? binder, object?[]
+            // Array construction is a special case.
+            if (DeclaringType is not null && DeclaringType.IsArray)
+            {
+                return InvokeArrayCtorWorker(arguments);
```
Before this change, the argument array reflected the coerced argument values. It does not seem to be the case anymore now that we are skipping the copy-out below.

E.g. this prints `System.Int32` today:
```csharp
using System.Globalization;
using System.Reflection;

var ctor = typeof(int[,]).GetConstructor(new[] { typeof(int), typeof(int) });
object[] arguments = new object[] { "1", "2" };
var arr = ctor.Invoke(BindingFlags.Default, new MyBinder(), arguments, null);
Console.WriteLine(arguments[0].GetType());

class MyBinder : Binder
{
    public override FieldInfo BindToField(BindingFlags bindingAttr, FieldInfo[] match, object value, CultureInfo? culture)
        => throw new NotImplementedException();

    public override MethodBase BindToMethod(BindingFlags bindingAttr, MethodBase[] match, ref object?[] args, ParameterModifier[]? modifiers, CultureInfo? culture, string[]? names, out object? state)
        => throw new NotImplementedException();

    public override object ChangeType(object value, Type type, CultureInfo? culture)
        => int.Parse((string)value);

    public override void ReorderArgumentArray(ref object?[] args, object state)
        => throw new NotImplementedException();

    public override MethodBase? SelectMethod(BindingFlags bindingAttr, MethodBase[] match, Type[] types, ParameterModifier[]? modifiers)
        => throw new NotImplementedException();

    public override PropertyInfo? SelectProperty(BindingFlags bindingAttr, PropertyInfo[] match, Type? returnType, Type[]? indexes, ParameterModifier[]? modifiers)
        => throw new NotImplementedException();
}
```

AaronRobinsonMSFT replied:
Bah. Wonderful consequence of this behavior. Fixed and added test.

AaronRobinsonMSFT committed:
coreclr/

Properly widen the input by respecting endianness.
Revert changes to InvokeUtil::CreatePrimitiveValue.

AaronRobinsonMSFT committed:
libraries/

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.cs:
```diff
@@ -165,7 +165,17 @@ public override object Invoke(BindingFlags invokeAttr, Binder? binder, object?[]
+            object retValue;
+
+            // Array construction is a special case.
+            if (DeclaringType is not null && DeclaringType.IsArray)
```
Is moving the array check here affect performance in measurable way?

AaronRobinsonMSFT replied:
I fixed the widening logic in the original path and reran the scenarios. I see very little impact on the change. Thoughts?

Old:
       230 ms: AllocateArray[100]: 1000000
       139 ms: DefaultConstructor: 1000000
       200 ms: IntConstructor: 1000000

New:
       228 ms: AllocateArray[100]: 1000000
       141 ms: DefaultConstructor: 1000000
       201 ms: IntConstructor: 1000000

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.CoreCLR.cs:
```diff
@@ -212,17 +212,24 @@ private void InvokeClassConstructor()
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private object InvokeCtorWorker(BindingFlags invokeAttr, in Span<object?> arguments)
```
Nit: You can also delete `in` on this one

jkotas commented on src/coreclr/vm/reflectioninvocation.cpp:
```diff
@@ -785,17 +747,6 @@ FCIMPL5(Object*, RuntimeMethodHandle::InvokeMethod,
-
         MethodTable * pMT = ownerType.AsMethodTable();
 
         {
```
We still have this special check for string constructors here. It makes me wonder whether it is an actual improvement to move the special check for array constructors to a different spot. I think it made more logical sense to have both these special checks together.

(I do not have a strong opinion on this.)

AaronRobinsonMSFT replied:
I'll move it back. The more I look into this less valuable this seems. I am going to consider this PR for fixing the widening logic and adding some much needed test coverage for Reflection scenarios.

jkotas reviewed (approved):
LGTM otherwise

AaronRobinsonMSFT committed:
coreclr/

Revert some non-impactful changes

AaronRobinsonMSFT committed:
libraries/

Revert some non-impactful changes

AaronRobinsonMSFT committed:
mono/

Revert some non-impactful changes

AaronRobinsonMSFT committed:
mono/

Remove additional changes.
